distribution/lib/Standard/Database/0.0.0-dev/src/Data/Take_Drop_Helpers.enso

```diff
@@ -0,0 +1,79 @@
+from Standard.Base import all
+
+import Standard.Base.Data.Index_Sub_Range as Index_Sub_Range_Module
+import Standard.Base.Errors.Illegal_Argument.Illegal_Argument
+from Standard.Base.Data.Index_Sub_Range import normalize_ranges, resolve_ranges, sort_and_merge_ranges
+
+from Standard.Table import Set_Mode
+
+import project.Data.Column.Column
+import project.Data.Table.Table
+from project.Errors import Unsupported_Database_Operation
+
+type Take_Drop
+    Take
+    Drop
+
+## PRIVATE
+   Apply `take` or `drop` to a table, returning the specified by the selector.
+take_drop_helper : Take_Drop -> Table -> (Index_Sub_Range | Range | Integer) -> Table
+take_drop_helper take_drop table selector =
+    check_supported selector <|
+        length = table.row_count
+        ranges = cleanup_ranges (collect_ranges take_drop length selector)
+
+        if ranges.is_empty then table.remove_all_rows else
+            # Filter on row column. Add the row column at the start, remove it at the end.
+            row_column_name = Table.make_temp_column_name
+            table_with_row_number = table.add_row_number name=row_column_name from=0
+
+            subqueries = ranges.map range->
+                generate_subquery table_with_row_number row_column_name range
+            combined = subqueries.reduce (a-> b-> a.union b)
+            combined.remove_columns row_column_name
```

Comment:
This looks concerning to me. Creating an union of so many sub-queries seems rather costly. I'm not sure the DB will be able to optimize that well enough, I imagine for `K` ranges we will have an union of `K` subqueries meaning `K` iterations through the whole result set.

IMO we should fold this into a single query with `OR` conditions. That will result in a single (albeit a bit more heavy) pass over the result set. IMO that should be faster and simpler. I really did not expect to see `union` in the implementation of `take`/`drop`.

Could we consider an alternative solution without `union`?

We can construct a single filter column and filter by it. We can 'merge' the ranges by `OR`ing the sub-filter conditions.

Something along the lines of:

```suggestion
            filters = ranges.map range->
                generate_filter row_column_name range
            combined_filter = filters.reduce (||)
            filtered = table_with_row_number.filter combined_filter
            filtered.remove_columns row_column_name
```

Reply:
My first implementation did not use union; it just built a single compound filter for all of the ranges. But then these tests failed:

```
            beta.take (Index_Sub_Range.By_Index [-1, -1]) . to_vector . should_equal ["H", "H"]
            alpha.take (Index_Sub_Range.By_Index [0, 0, 3.up_to 100]) . to_vector . should_equal [1, 1, 4, 5, 6, 7, 8]
```

The output of these `take` operations includes duplicates, and is also order-sensitive. If you use `By_Index` with a repeated index (such as -1 or 0 here), then with just a compound filter, you'll have something like `Filter_Condition.Or (Filter_Condition.Equal 0) (Filter_Condition.Equal 0)`, which only produces one copy of the row, not two. As you can see, the test output `[1, 1, 4, 5, 6, 7, 8]` expects two copies of `1`.

In addition, a `By_Index` argument might have multiple ranges, possibly overlapping. Combining everything into a single filter will not respect the order of the ranges within the `By_Index` field.

So under the assumption that the tests are specifying exactly what we expect from take/drop, I switched from a filter to union. I have another task to consider some optimizations, and for that task I can consider combining multiple selects into a single select with a compound filter, in the case that that won't change the ordering.

Or, we can consider whether preserving duplicates is essential to this operation. @jdunkerley thoughts?

Reply:
Oh I see! That makes total sense, I forgot about this edge case.

Then the approach looks good. Let's note that for most cases we will have a _single_ range anyway (`take (First 5)` etc. will be most common I think) - so there will be no union anyway.